### ISSUE TITLE
Mypy points wrong "return" in async gens, pyright misses it

### ISSUE THREAD

**Issue by betafcc** [microsoft/pyright#4027]
**Describe the bug**
Got some inconsistent behaviour between mypy and pyright when working with async gens, both when defining and when consuming them.

**To Reproduce**

Using pyright 1.1.274 vs mypy 0.982 

#### 1 - "return" with value in async generator is not allowed

This is fine in pyright:

```python
from typing import TypeVar, AsyncIterable

A = TypeVar("A")


async def drop(n: int, fa: AsyncIterable[A]) -> AsyncIterable[A]:
    if n <= 0:
        return fa  # <----- the subject

    g = aiter(fa)

    while n > 0:
        try:
            await g.__anext__()
        except StopAsyncIteration:
            return
        n -= 1

    while True:
        try:
            yield await g.__anext__()
        except StopAsyncIteration:
            return
```

But mypy reports

```sh
❱ mypy dirty_bits/helpers/aitertools.py 
dirty_bits/helpers/aitertools.py:100: error: "return" with value in async generator is not allowed
```

#### 2 - Missing return statement

considered putting in another issue but not sure if worth since pyright seem to be working as I expected here:

```python
from typing import AsyncIterable, TypeVar

A = TypeVar("A")

async def first(fa: AsyncIterable[A]) -> A | None:
    async for a in fa:
        return a
```

Pyright issues no warning but mypy reports:

```sh
❱ mypy dirty_bits/helpers/aitertools.py 
dirty_bits/helpers/aitertools.py:127: error: Missing return statement
```

Do note I'm explicitly annotating with `A | None`, as I was expecting the possibility of not returning, so, probably this one is on mypy right?

**Comment by betafcc** [comment]
Just noticed the same occurs in non-async Iterable:

```python
from __future__ import annotations

from typing import Iterable, Iterator, Generic, TypeVar

A = TypeVar("A")


class Iter(Generic[A]):
    def __init__(self, fa: Iterable[A]) -> None:
        self.fa = fa

    def __iter__(self) -> Iterator[A]:
        return self.fa.__iter__()

    def enumerate(self) -> Iter[tuple[int, A]]:
        def gen():
            i = -1
            for a in self:
                i += 1
                yield i, a

        return Iter(gen())

    def wrong(self) -> Iterable[A]:
        for i, value in self.enumerate():
            reveal_type(i)  # Type of "i" is "int"
            reveal_type(value)  # Type of "value" is "tuple[int, A@Iter]"
            yield value  # Expression of type "tuple[int, A@Iter]" cannot be assigned to yield type "A@Iter"


def right(fa: Iter[A]) -> Iterable[A]:
    for i, value in fa.enumerate():
        reveal_type(i)  # Type of "i" is "int"
        reveal_type(value)  # Type of "value" is "A@right"
        yield value  # ok, type A as annotated
```

**Comment by erictraut** [comment]
On item 1, I agree this is something that pyright should detect. I've added support for reporting an error if an async generator includes a return statement with an explicit return value. This will be included in the next release.

On item 2, this isn't a bug in pyright. You are implicitly returning `None` from this function, and the return type is annotated as `A | None`, so it is sound from a type checking standpoint. Mypy has somewhat different behaviors when it comes to implicit return values. I don't agree with mypy's behavior in this case, but I don't think the mypy maintainers would consider it a bug.

On the other item you filed, I'm not sure what you mean. In any case, it appears to be unrelated to this enhancement request. If there is a separate enhancement request or bug that you'd like to report, please open another issue.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.275, which I just published. It will also be included in a future release of pylance.
